### DISCUSSION TITLE
Guidance on debugging Catalyst

### DISCUSSION THREAD

**Discussion by jmahony** [bigcommerce/catalyst#1666] [Q&A] [answered]
Hello!

Ever since the `--turbo` flag was added with this [commit](https://github.com/bigcommerce/catalyst/commit/16e3a763571324dccd9031a79e400409eff9ee0c#diff-7f67769260741e2563407af949f7e54fbf0431d1c607933f6e8a8094e3219e26L7) our team has been having issues debugging the project. Breakpoints fail to bind in the `core` project, other packages such as `client` work as expected. The only workaround we've found is to remove the `--turbo` flag which doesn't seem ideal.

We're using VSCode via WSL on Windows 11.

Our previously working `launch.json` as per the [Next.js docs](https://nextjs.org/docs/app/building-your-application/configuring/debugging):
```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Next.js: debug server-side",
      "type": "node-terminal",
      "request": "launch",
      "command": "pnpm run dev",
      "cwd": "${workspaceFolder}/core"
    },
    {
      "name": "Next.js: debug client-side",
      "type": "chrome",
      "request": "launch",
      "url": "http://localhost:3000"
    },
    {
      "name": "Next.js: debug client-side (Firefox)",
      "type": "firefox",
      "request": "launch",
      "url": "http://localhost:3000",
      "reAttach": true,
      "pathMappings": [
        {
          "url": "webpack://_N_E",
          "path": "${workspaceFolder}"
        }
      ]
    },
    {
      "name": "Next.js: debug full stack",
      "type": "node",
      "request": "launch",
      "program": "${workspaceFolder}/node_modules/.bin/next",
      "runtimeArgs": ["--inspect"],
      "skipFiles": ["<node_internals>/**"],
      "serverReadyAction": {
        "action": "debugWithEdge",
        "killOnServerStop": true,
        "pattern": "- Local:.+(https?://.+)",
        "uriFormat": "%s",
        "webRoot": "${workspaceFolder}"
      }
    }
  ]
}
```

Any guidance would be appreciated.

**Comment by deini** [comment]
This seems related, are you seeing a similar error?

https://github.com/vercel/next.js/issues/72789

Is it not working on any of those profiles (ie: server-side / full stack)? Or are some of them working?

**Reply by jmahony** [reply]
Thank you for your reply :) 
# Full stack

To be honest I've not really tried to use full stack, but it doesn't work at all. With the config I provided above I get the error:
```
Uncaught Error Error: Cannot find module '/home/josh/projects/catalyst/node_modules/.bin/next'
    at Module._resolveFilename (internal/modules/cjs/loader:1225:15)
    at Module._load (internal/modules/cjs/loader:1051:27)
    at executeUserEntryPoint (internal/modules/run_main:174:12)
    at <anonymous> (internal/main/run_main_module:28:49)
```
If I update the `program` field with `"${workspaceFolder}/core/node_modules/.bin/next"` I then get the error:

```
Process exited with code 1
Uncaught SyntaxError /home/josh/projects/catalyst/core/node_modules/.bin/next:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at wrapSafe (internal/modules/cjs/loader:1378:20)
    at Module._compile (internal/modules/cjs/loader:1428:41)
    at Module._extensions..js (internal/modules/cjs/loader:1548:10)
    at Module.load (internal/modules/cjs/loader:1288:32)
    at Module._load (internal/modules/cjs/loader:1104:12)
    at executeUserEntryPoint (internal/modules/run_main:174:12)
    at <anonymous> (internal/main/run_main_module:28:49)
```

# Server-side
With server-side, I don't appear to get any errors in the debug console like in the link you've provided:
![Z03kgNmhop](https://github.com/user-attachments/assets/fecb6087-3b6b-44bf-9311-663ee2bf4dc0)

It just looks like the source maps are loaded when using the `--turbo` flag. 

![Cursor_rrSdjHZSzq](https://github.com/user-attachments/assets/83d74b3f-9c0d-414e-a4f5-c25ad6f361dc)

As a note, I'm testing this with `catalyst/main` to ensure it has nothing to do with our changes.

I should also say client-side debugging in Chrome is working perfectly, this is just server-side. My issue sounds similar to this: https://github.com/vercel/next.js/issues/62008, but isn't isolated to server components.

**Comment by mttsdlerrdgwy** [comment]
Also hitting this issue with VSCode on Windows. Updating turbo to 2.3.3 doesn't help.

Turbopack is also causing breakpoints to be unbound in the vanilla Next.Js app `npx create-next-app@latest`. 
They work there when `--turbo` removed from the dev script, but it doesn't seem as easy to rectify in Catalyst.

**Comment by chanceaclark** [comment, marked as answer]
We ended up removing the `--turbo` flag as it does cause issues, not just with debugging. I think this is the path forward until the Next.js team releases the flag as stable.
